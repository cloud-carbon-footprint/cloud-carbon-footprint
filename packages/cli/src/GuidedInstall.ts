/*
 * © 2020 ThoughtWorks, Inc. All rights reserved.
 */

import { confirm, list, prompt } from 'typed-prompts'
import { prop } from 'ramda'

const log = (m: string) => console.log(`\n# ${m}\n`)
const listPrompt = (message: string, options: string[]) =>
  prompt<{ key: string }>([list('key', message, options)]).then(prop('key'))
const confirmPrompt = (message: string) =>
  prompt<{ key: string }>([confirm('key', message + '\n')]).then((result) => {
    if (!result.key) {
      log('Please try again when you have completed this step.')
      process.exit(0)
    }
    console.log()
  })

async function GuidedInstall() {
  log(
    `Welcome to CCF's Guided Install. This tool will walk you through the documentation and automate certain parts of it. Please see https://github.com/ThoughtWorks-Cleantech/cloud-carbon-footprint/tree/trunk/docs for details.`,
  )

  const provider: string = await listPrompt(
    'Choose a cloud provider:',
    Object.keys(providers),
  )

  log(`${provider} setup`)
  const setup = providers[provider]
  await setup()
}

const providers: { [key: string]: () => Promise<void> } = {
  AWS: AWSSetup,
  GCP: () => null,
  Azure: () => null,
}

async function AWSSetup() {
  // ideally the docs live on the microsite and are generated from a json file that this also reads
  // for now just manually copying the docs
  // some of these steps seem like they could be achieved with terraform / provider specific config files

  await confirmPrompt(
    `Ensure your aws account has the correct permissions.\n\t- You will need an [IAM user](https://aws.amazon.com/premiumsupport/knowledge-center/create-access-key/) that can create access-keys and modify your billing settings.`,
  )

  await confirmPrompt(
    `Enable the Cost and Usage Reports Billing AWS feature.\n\t- This feature needs to be enabled so your account can start generating cost and usage reports.\n\t- To enable, navigate to your account's billing section, and click on the "Cost and Usage Reports" tab.\n\t- Make sure to select “Amazon Athena” for report data integration.\n\t- Reference Cost and Usage Reports documentation [here](https://docs.aws.amazon.com/cur/latest/userguide/what-is-cur.html).`,
  )

  await confirmPrompt(
    `Setup Athena DB to save the Cost and Usage Reports.\n\t- In addition to generating reports, we use Athena DB to save the details of those reports in a DB, so we can run queries on them.\n\t- This is a standard AWS integration, outlined [here](https://docs.aws.amazon.com/cur/latest/userguide/cur-query-athena.html)`
  )

  await confirmPrompt(
    `Configure aws credentials locally, using awscli.\n\t- After [installing awscli](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-install.html), run \`aws configure\` and provide your access key and secret access key.\n\t- Also make sure you select the same region as the one you created your cost and usage reports in.`
  )

  log('TODO create env file')

  log('')
}

GuidedInstall().catch((error) => {
  console.error(`Something went wrong: ${error.message}`)
})
