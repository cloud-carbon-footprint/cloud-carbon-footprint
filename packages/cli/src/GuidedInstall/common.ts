/*
 * Â© 2020 ThoughtWorks, Inc. All rights reserved.
 */

import { promisify } from 'util'
import { exec as execCb } from 'child_process'
import dotenv from 'dotenv'
import fs from 'fs-extra'
import { confirm, input, list, prompt } from 'typed-prompts'
import { prop } from 'ramda'

const exec = promisify(execCb)
const stat = promisify(fs.stat)

export const log = (m: string, leading = true): void =>
  console.log(`${leading ? '\n' : ''}# ${m}\n`)

export const lineBreak = (): void => console.log()

export const listPrompt = (
  message: string,
  options: string[],
): Promise<string> =>
  prompt<{ key: string }>([list('key', message, options)]).then(prop('key'))

export const confirmPrompt = (
  message: string,
  question = 'Is this step complete?',
  requireYes = true,
): Promise<boolean> =>
  prompt<{ key: boolean }>([confirm('key', message + `\n${question}`)])
    .then(prop('key'))
    .then((result) => {
      if (requireYes && !result) {
        log('Please try again when you have completed this step.')
        process.exit(0)
      }
      console.log()
      return result
    })

export const inputPrompt = (
  message: string,
  required = true,
): Promise<string> =>
  prompt<{ key: string }>([input('key', message)])
    .then(prop('key'))
    .then((result: string) => {
      if (required && !result) {
        log('Please enter a value.')
        return inputPrompt(message, required)
      }
      return result
    })

export const runCmd = async (cmd: string): void => {
  try {
    await exec(cmd)
  } catch (error) {
    process.stdout.write(error.stderr)
    process.stdout.write(error.stdout)
    throw new Error(`Could not execute command ${cmd}`)
  }
}
// will load existing .env and overwrite the given keys
export const createEnvFile = async (
  dir: string,
  env: { [key: string]: string },
): void => {
  const path = `${dir}.env`
  log(`Creating ${path}...`, false)
  const exists = await stat(path)
    .then(() => true)
    .catch(() => false)

  const existing = exists ? dotenv.config({ path }) : {}
  if (existing.error) {
    throw existing.error
  }

  const newEnv = { ...existing.parsed, ...env }
  await fs.writeFile(
    path,
    Object.entries(newEnv).reduce(
      (config, [k, v]) => `${config}${k}=${v}\n`,
      '',
    ),
  )
  log('...done', false)
}
