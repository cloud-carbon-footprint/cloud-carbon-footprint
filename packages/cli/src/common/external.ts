/*
 * Â© 2021 Thoughtworks, Inc.
 */

/* istanbul ignore file */
import { exec as _exec } from 'child_process'
import fs from 'fs-extra'
import { promisify } from 'util'
import { prompt as _prompt, Question } from 'typed-prompts'
import { resolve as _resolve } from 'path'
import dotenv from 'dotenv'

export const exec = promisify(_exec)

export const stat = promisify(fs.stat)

export const writeFile = fs.writeFile

export const log = (m: string, leading = true): void =>
  console.log(`${leading ? '\n' : ''}# ${m}\n`)

export const lineBreak = (): void => console.log()

export const exit = (code = 0): void => process.exit(code)

export const prompt = <T>(question: Question): Promise<T> =>
  _prompt<T>([question])

export const readConfig = (
  path: string,
): { parsed?: { [key: string]: string }; error?: Error } =>
  dotenv.config({ path })

export const resolve = _resolve

export const runCmd = async (cmd: string): Promise<void> => {
  try {
    await exec(cmd)
  } catch (error) {
    process.stdout.write(error.stderr)
    process.stdout.write(error.stdout)
    throw new Error(`Could not execute command ${cmd}`)
  }
}
