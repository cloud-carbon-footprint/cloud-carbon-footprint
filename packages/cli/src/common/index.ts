/*
 * Â© 2021 Thoughtworks, Inc.
 */

import { confirm, input, list } from 'typed-prompts'
import { prop } from 'ramda'
import {
  exit,
  lineBreak,
  log,
  prompt,
  readConfig,
  resolve,
  stat,
  writeFile,
} from './external'

export type EnvConfig = { [key: string]: string }

export const microsite = 'https://cloud-carbon-footprint.org'

export const listPrompt = (
  message: string,
  options: string[],
): Promise<string> =>
  prompt<{ key: string }>(list('key', message, options)).then(prop('key'))

export const confirmPrompt = (
  message: string,
  question = 'Is this step complete?',
  requireYes = true,
): Promise<boolean> =>
  prompt<{ key: boolean }>(confirm('key', message + `\n${question}`))
    .then(prop('key'))
    .then((result) => {
      if (requireYes && !result) {
        log('Please try again when you have completed this step.')
        exit()
      }
      lineBreak()
      return result
    })

export const inputPrompt = (
  message: string,
  required = true,
): Promise<string> =>
  prompt<{ key: string }>(input('key', message))
    .then(prop('key'))
    .then((result: string) => {
      if (required && !result) {
        log('Please enter a value.')
        return inputPrompt(message, required)
      }
      return result
    })

export const createEnvFile = async (
  dir: string,
  env: { [key: string]: string },
): Promise<void> => {
  const path = resolve(`${dir}.env`)
  log(`Creating/updating ${path}...`, false)
  const exists = await stat(path)
    .then(() => true)
    .catch(() => false)

  const existing = exists ? readConfig(path) : { parsed: undefined }
  if (existing.error) {
    throw existing.error
  }

  const newEnv = { ...existing.parsed, ...env }
  await writeFile(
    path,
    Object.entries(newEnv).reduce(
      (config, [k, v]) => `${config}${k}=${v}\n`,
      '',
    ),
  )
  log('...done', false)
}
