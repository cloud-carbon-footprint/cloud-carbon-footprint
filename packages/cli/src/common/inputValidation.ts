/*
 * Â© 2021 Thoughtworks, Inc.
 */
import {
  EstimationRequestValidationError,
  OnPremiseDataInput,
} from '@cloud-carbon-footprint/common'

type cliDataInput = {
  groupBy: string
}

export const validateOnPremiseInput = (inputData: OnPremiseDataInput[]) => {
  inputData.map((inputRow: OnPremiseDataInput) => {
    if (
      !inputRow.cpuDescription ||
      !inputRow.machineType ||
      !inputRow.memory ||
      !inputRow.startTime ||
      !inputRow.endTime ||
      !inputRow.dailyUptime ||
      !inputRow.weeklyUptime ||
      !inputRow.monthlyUptime ||
      !inputRow.annualUptime
    ) {
      throw new Error(
        'Input data is incorrect. Please check your input data file and try again.',
      )
    }
  })
}

export const validateCliInput = (inputData: cliDataInput) => {
  const groupingOptions = ['day', 'dayAndService', 'service']
  if (!groupingOptions.includes(inputData.groupBy)) {
    throw new EstimationRequestValidationError(
      `GroupBy param is incorrect. Please specify one of the following grouping methods: ${groupingOptions.join(
        ' | ',
      )}`,
    )
  }
}
