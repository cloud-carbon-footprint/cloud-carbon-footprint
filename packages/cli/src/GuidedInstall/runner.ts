/*
 * Â© 2021 Thoughtworks, Inc.
 */

import { createEnvFile, EnvConfig, listPrompt, microsite } from '../common'
import { log } from '../common/external'
import { AWSSetup } from './aws'
import { GCPSetup } from './gcp'
import { AzureSetup } from './azure'

type ProvidersConfig = {
  [key: string]: { setup: () => Promise<EnvConfig>; docs: string }
}
const providers: ProvidersConfig = {
  AWS: {
    setup: AWSSetup,
    docs: '/docs/aws',
  },
  GCP: {
    setup: GCPSetup,
    docs: '/docs/gcp',
  },
  Azure: {
    setup: AzureSetup,
    docs: '/docs/azure',
  },
}

export async function GuidedInstall(): Promise<void> {
  log(
    `Please follow the following prompts to configure Cloud Carbon Footprint to connect to a cloud provider. To connect to multiple cloud providers, please repeat this script for each additional provider. For an understanding of the default configurations and other configuration options, please visit the docs on the microsite: ${microsite}/docs/introduction`,
  )

  const provider: string = await listPrompt(
    'Choose a cloud provider to configure:',
    Object.keys(providers),
  )

  const { setup, docs } = providers[provider]
  log(`${provider} setup. See ${microsite}${docs} for additional details.`)
  const env = await setup()

  await createEnvFile('./', env)
  await createEnvFile('../api/', env)

  log(
    `Your Cloud Carbon Footprint application is now configured to use your cloud provider data. You can see these settings in the .env files listed above. To connect to an additional cloud provider or update an existing one, please re-run this script. For questions or more configuration options, please visit ${microsite}/docs/introduction`,
  )
  log(
    `Please make sure your appConfig object in the client Config file (packages/client/src/Config.ts) includes your cloud provider(s) in the CURRENT_PROVIDERS array`,
  )
}
