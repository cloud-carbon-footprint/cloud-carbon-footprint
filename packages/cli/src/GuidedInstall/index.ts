/*
 * Â© 2020 ThoughtWorks, Inc. All rights reserved.
 */

/* istanbul ignore file */
import { GuidedInstall } from './runner'

GuidedInstall().catch((error) => {
  console.error(`Something went wrong: ${error.message}`)
})
