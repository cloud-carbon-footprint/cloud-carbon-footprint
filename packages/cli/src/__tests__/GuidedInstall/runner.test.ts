/*
 * Â© 2021 Thoughtworks, Inc.
 */

import * as common from '../../common'
import * as external from '../../common/external'
import { GuidedInstall } from '../../GuidedInstall/runner'

jest.mock('../../common')
const mockListPrompt: jest.Mock = common.listPrompt as jest.Mock
const mockConfirmPrompt: jest.Mock = common.confirmPrompt as jest.Mock
const mockInputPrompt: jest.Mock = common.inputPrompt as jest.Mock
const mockCreateEnvFile: jest.Mock = common.createEnvFile as jest.Mock

jest.mock('../../common/external')
const mockLog: jest.Mock = external.log as jest.Mock
const mockRunCmd: jest.Mock = external.runCmd as jest.Mock

describe('GuidedInstall', () => {
  afterEach(() => {
    jest.resetAllMocks()
  })

  describe('Azure', () => {
    beforeEach(() => {
      mockListPrompt.mockResolvedValue('Azure')
      mockConfirmPrompt.mockResolvedValue(true)
      mockInputPrompt
        .mockResolvedValueOnce('client id')
        .mockResolvedValueOnce('client secret')
        .mockResolvedValueOnce('tenant id')
    })

    it('creates an env file for cli and api with the given config', async () => {
      await GuidedInstall()

      const expectedEnv = {
        AZURE_CLIENT_ID: 'client id',
        AZURE_CLIENT_SECRET: 'client secret',
        AZURE_TENANT_ID: 'tenant id',
        AZURE_USE_BILLING_DATA: 'true',
      }
      expect(mockCreateEnvFile).toHaveBeenCalledWith('./', expectedEnv)
      expect(mockCreateEnvFile).toHaveBeenCalledWith('../api/', expectedEnv)
    })

    it('shows the correct confirmation prompts for manual instructions', async () => {
      await GuidedInstall()

      expect(mockConfirmPrompt.mock.calls).toMatchSnapshot()
    })

    it('shows the correct input prompts for collecting env vars', async () => {
      await GuidedInstall()

      expect(mockInputPrompt.mock.calls).toMatchSnapshot()
    })

    it('shows the correct informational messages and documentation references', async () => {
      await GuidedInstall()

      expect(mockLog.mock.calls).toMatchSnapshot()
    })
  })

  describe('GCP', () => {
    beforeEach(() => {
      mockListPrompt.mockResolvedValue('GCP')
      mockConfirmPrompt.mockResolvedValue(true)
      mockInputPrompt
        .mockResolvedValueOnce('private key file path')
        .mockResolvedValueOnce('account id')
        .mockResolvedValueOnce('account name')
        .mockResolvedValueOnce('BIG table')
    })

    it('creates an env file for cli and api with the given config', async () => {
      await GuidedInstall()

      const expectedEnv = {
        GCP_BILLING_PROJECT_ID: 'account id',
        GCP_BILLING_PROJECT_NAME: 'account name',
        GCP_BIG_QUERY_TABLE: 'BIG table',
        GCP_USE_BILLING_DATA: 'true',
        GCP_USE_CARBON_FREE_ENERGY_PERCENTAGE: 'true',
        GOOGLE_APPLICATION_CREDENTIALS: 'private key file path',
      }
      expect(mockCreateEnvFile).toHaveBeenCalledWith('./', expectedEnv)
      expect(mockCreateEnvFile).toHaveBeenCalledWith('../api/', expectedEnv)
    })

    it('shows the correct confirmation prompts for manual instructions', async () => {
      await GuidedInstall()

      expect(mockConfirmPrompt.mock.calls).toMatchSnapshot()
    })

    it('shows the correct input prompts for collecting env vars', async () => {
      await GuidedInstall()

      expect(mockInputPrompt.mock.calls).toMatchSnapshot()
    })

    it('shows the correct informational messages and documentation references', async () => {
      await GuidedInstall()

      expect(mockLog.mock.calls).toMatchSnapshot()
    })
  })

  describe('AWS', () => {
    describe('with awscli config', () => {
      beforeEach(() => {
        mockListPrompt.mockResolvedValue('AWS')
        mockConfirmPrompt
          .mockResolvedValueOnce(true)
          .mockResolvedValueOnce(true)
          .mockResolvedValueOnce(true)
          .mockResolvedValueOnce(true)
          .mockResolvedValueOnce(true)
        mockInputPrompt
          .mockResolvedValueOnce('account id')
          .mockResolvedValueOnce('account name')
          .mockResolvedValueOnce('region')
          .mockResolvedValueOnce('role name')
          .mockResolvedValueOnce('athena db name')
          .mockResolvedValueOnce('athena db table')
          .mockResolvedValueOnce('athena query result location')
          .mockResolvedValueOnce('access key id')
          .mockResolvedValueOnce('secret access key')
          .mockResolvedValueOnce('profile')
      })

      it('creates an env file for cli and api with the given config', async () => {
        await GuidedInstall()

        const expectedEnv = {
          AWS_ATHENA_DB_NAME: 'athena db name',
          AWS_ATHENA_DB_TABLE: 'athena db table',
          AWS_ATHENA_QUERY_RESULT_LOCATION: 'athena query result location',
          AWS_ATHENA_REGION: 'region',
          AWS_BILLING_ACCOUNT_ID: 'account id',
          AWS_BILLING_ACCOUNT_NAME: 'account name',
          AWS_TARGET_ACCOUNT_ROLE_NAME: 'role name',
          AWS_USE_BILLING_DATA: 'true',
        }
        expect(mockCreateEnvFile).toHaveBeenCalledWith('./', expectedEnv)
        expect(mockCreateEnvFile).toHaveBeenCalledWith('../api/', expectedEnv)
      })

      it('shows the correct confirmation prompts for manual instructions', async () => {
        await GuidedInstall()

        expect(mockConfirmPrompt.mock.calls).toMatchSnapshot()
      })

      it('shows the correct input prompts for collecting env vars', async () => {
        await GuidedInstall()

        expect(mockInputPrompt.mock.calls).toMatchSnapshot()
      })

      it('runs `aws configure` commands', async () => {
        await GuidedInstall()

        expect(mockRunCmd.mock.calls).toMatchSnapshot()
      })

      it('shows the correct informational messages and documentation references', async () => {
        await GuidedInstall()

        expect(mockLog.mock.calls).toMatchSnapshot()
      })
    })

    describe('without awscli config', () => {
      beforeEach(() => {
        mockListPrompt.mockResolvedValue('AWS')
        mockConfirmPrompt
          .mockResolvedValueOnce(true)
          .mockResolvedValueOnce(true)
          .mockResolvedValueOnce(true)
          .mockResolvedValueOnce(true)
          .mockResolvedValueOnce(false)
          .mockResolvedValueOnce(true)
        mockInputPrompt
          .mockResolvedValueOnce('account id')
          .mockResolvedValueOnce('account name')
          .mockResolvedValueOnce('region')
          .mockResolvedValueOnce('role name')
          .mockResolvedValueOnce('athena db name')
          .mockResolvedValueOnce('athena db table')
          .mockResolvedValueOnce('athena query result location')
      })

      it('creates an env file for cli and api with the given config', async () => {
        await GuidedInstall()

        const expectedEnv = {
          AWS_ATHENA_DB_NAME: 'athena db name',
          AWS_ATHENA_DB_TABLE: 'athena db table',
          AWS_ATHENA_QUERY_RESULT_LOCATION: 'athena query result location',
          AWS_ATHENA_REGION: 'region',
          AWS_BILLING_ACCOUNT_ID: 'account id',
          AWS_BILLING_ACCOUNT_NAME: 'account name',
          AWS_TARGET_ACCOUNT_ROLE_NAME: 'role name',
          AWS_USE_BILLING_DATA: 'true',
        }
        expect(mockCreateEnvFile).toHaveBeenCalledWith('./', expectedEnv)
        expect(mockCreateEnvFile).toHaveBeenCalledWith('../api/', expectedEnv)
      })

      it('shows the correct confirmation prompts for manual instructions', async () => {
        await GuidedInstall()

        expect(mockConfirmPrompt.mock.calls).toMatchSnapshot()
      })

      it('shows the correct input prompts for collecting env vars', async () => {
        await GuidedInstall()

        expect(mockInputPrompt.mock.calls).toMatchSnapshot()
      })

      it('does not run `aws configure` commands', async () => {
        await GuidedInstall()

        expect(mockRunCmd).not.toHaveBeenCalled()
      })

      it('shows the correct informational messages and documentation references', async () => {
        await GuidedInstall()

        expect(mockLog.mock.calls).toMatchSnapshot()
      })
    })
  })

  describe('MongoDB', () => {
    describe('with auth certificate', () => {
      beforeEach(() => {
        mockListPrompt.mockResolvedValue('MongoDB')
        mockConfirmPrompt
          .mockResolvedValueOnce(true)
          .mockResolvedValueOnce(true)
          .mockResolvedValueOnce(true)
        mockInputPrompt.mockResolvedValueOnce('mongodb://localhost:27017')
        mockInputPrompt.mockResolvedValueOnce('~/keys/mongodb-certificate.pem')
      })

      it('creates an env file for cli and api with the given config', async () => {
        await GuidedInstall()

        const expectedEnv = {
          CACHE_MODE: 'MONGODB',
          MONGODB_URI: 'mongodb://localhost:27017',
          MONGODB_CREDENTIALS: '~/keys/mongodb-certificate.pem',
        }

        expect(mockCreateEnvFile).toHaveBeenCalledWith('./', expectedEnv)
        expect(mockCreateEnvFile).toHaveBeenCalledWith('../api/', expectedEnv)
      })
    })
  })

  describe('without auth certificate', () => {
    beforeEach(() => {
      mockListPrompt.mockResolvedValue('MongoDB')
      mockConfirmPrompt
        .mockResolvedValueOnce(true)
        .mockResolvedValueOnce(true)
        .mockResolvedValueOnce(false)
      mockInputPrompt.mockResolvedValueOnce('mongodb://localhost:27017')
    })

    it('creates an env file for cli and api with the given config', async () => {
      await GuidedInstall()

      const expectedEnv = {
        CACHE_MODE: 'MONGODB',
        MONGODB_URI: 'mongodb://localhost:27017',
      }

      expect(mockCreateEnvFile).toHaveBeenCalledWith('./', expectedEnv)
      expect(mockCreateEnvFile).toHaveBeenCalledWith('../api/', expectedEnv)
    })
  })
})
