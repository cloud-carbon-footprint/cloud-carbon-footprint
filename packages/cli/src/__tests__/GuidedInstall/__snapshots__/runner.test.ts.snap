// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GuidedInstall AWS with awscli config runs \`aws configure\` commands 1`] = `
Array [
  Array [
    "aws configure set aws_access_key_id access key id --profile profile",
  ],
  Array [
    "aws configure set aws_secret_access_key secret access key --profile profile",
  ],
  Array [
    "aws configure set region region --profile profile",
  ],
]
`;

exports[`GuidedInstall AWS with awscli config shows the correct confirmation prompts for manual instructions 1`] = `
Array [
  Array [
    "Ensure your AWS account has the correct permissions.
	- You will need an [IAM user](https://aws.amazon.com/premiumsupport/knowledge-center/create-access-key/) that can create access-keys and modify your billing settings.",
  ],
  Array [
    "Enable the Cost and Usage Reports Billing AWS feature.
	- This feature needs to be enabled so your account can start generating cost and usage reports.
	- To enable, navigate to your account's billing section, and click on the \\"Cost and Usage Reports\\" tab.
	- Make sure to select “Amazon Athena” for report data integration.
	- Reference Cost and Usage Reports documentation [here](https://docs.aws.amazon.com/cur/latest/userguide/what-is-cur.html).",
  ],
  Array [
    "Setup Athena DB to save the Cost and Usage Reports.
	- In addition to generating reports, we use Athena DB to save the details of those reports in a DB, so we can run queries on them.
	- This is a standard AWS integration, outlined [here](https://docs.aws.amazon.com/cur/latest/userguide/cur-query-athena.html)",
  ],
  Array [
    "Install the AWS CLI https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-install.html",
  ],
  Array [
    "Configure aws credentials locally, using awscli. See the documentation for more details on authenticating to AWS.",
    "Would you like this tool to configure your awscli? If no, you will need to do it yourself.",
    false,
  ],
]
`;

exports[`GuidedInstall AWS with awscli config shows the correct informational messages and documentation references 1`] = `
Array [
  Array [
    "Please follow the following prompts to configure Cloud Carbon Footprint to connect to a cloud provider. To connect to multiple cloud providers, please repeat this script for each additional provider. For an understanding of the default configurations and other configuration options, please visit the docs on the microsite: https://cloud-carbon-footprint.org/docs/introduction",
  ],
  Array [
    "AWS setup. See https://cloud-carbon-footprint.org/docs/aws for additional details.",
  ],
  Array [
    "Configuring AWS CLI...",
  ],
  Array [
    "...done",
    false,
  ],
  Array [
    "Your Cloud Carbon Footprint application is now configured to use your cloud provider data. You can see these settings in the .env files listed above. To connect to an additional cloud provider or update an existing one, please re-run this script. For questions or more configuration options, please visit https://cloud-carbon-footprint.org/docs/introduction",
  ],
  Array [
    "Please make sure your appConfig object in the client Config file (packages/client/src/Config.ts) includes your cloud provider(s) in the CURRENT_PROVIDERS array",
  ],
]
`;

exports[`GuidedInstall AWS with awscli config shows the correct input prompts for collecting env vars 1`] = `
Array [
  Array [
    "Enter AWS account id:",
  ],
  Array [
    "Enter AWS account name:",
  ],
  Array [
    "Enter AWS region:",
  ],
  Array [
    "Enter AWS role name:",
  ],
  Array [
    "Enter Athena DB name:",
  ],
  Array [
    "Enter Athena DB table:",
  ],
  Array [
    "Enter Athena DB query result location (eg s3://your-athena-query-results-location):",
  ],
  Array [
    "Enter AWS access key id:",
  ],
  Array [
    "Enter AWS secret access key:",
  ],
  Array [
    "Optionally enter AWS profile:",
    false,
  ],
]
`;

exports[`GuidedInstall AWS without awscli config shows the correct confirmation prompts for manual instructions 1`] = `
Array [
  Array [
    "Ensure your AWS account has the correct permissions.
	- You will need an [IAM user](https://aws.amazon.com/premiumsupport/knowledge-center/create-access-key/) that can create access-keys and modify your billing settings.",
  ],
  Array [
    "Enable the Cost and Usage Reports Billing AWS feature.
	- This feature needs to be enabled so your account can start generating cost and usage reports.
	- To enable, navigate to your account's billing section, and click on the \\"Cost and Usage Reports\\" tab.
	- Make sure to select “Amazon Athena” for report data integration.
	- Reference Cost and Usage Reports documentation [here](https://docs.aws.amazon.com/cur/latest/userguide/what-is-cur.html).",
  ],
  Array [
    "Setup Athena DB to save the Cost and Usage Reports.
	- In addition to generating reports, we use Athena DB to save the details of those reports in a DB, so we can run queries on them.
	- This is a standard AWS integration, outlined [here](https://docs.aws.amazon.com/cur/latest/userguide/cur-query-athena.html)",
  ],
  Array [
    "Install the AWS CLI https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-install.html",
  ],
  Array [
    "Configure aws credentials locally, using awscli. See the documentation for more details on authenticating to AWS.",
    "Would you like this tool to configure your awscli? If no, you will need to do it yourself.",
    false,
  ],
  Array [
    "Run \`aws configure\` and provide your access key and secret access key.
	- Also make sure you select the same region as the one you created your cost and usage reports in.",
  ],
]
`;

exports[`GuidedInstall AWS without awscli config shows the correct informational messages and documentation references 1`] = `
Array [
  Array [
    "Please follow the following prompts to configure Cloud Carbon Footprint to connect to a cloud provider. To connect to multiple cloud providers, please repeat this script for each additional provider. For an understanding of the default configurations and other configuration options, please visit the docs on the microsite: https://cloud-carbon-footprint.org/docs/introduction",
  ],
  Array [
    "AWS setup. See https://cloud-carbon-footprint.org/docs/aws for additional details.",
  ],
  Array [
    "Your Cloud Carbon Footprint application is now configured to use your cloud provider data. You can see these settings in the .env files listed above. To connect to an additional cloud provider or update an existing one, please re-run this script. For questions or more configuration options, please visit https://cloud-carbon-footprint.org/docs/introduction",
  ],
  Array [
    "Please make sure your appConfig object in the client Config file (packages/client/src/Config.ts) includes your cloud provider(s) in the CURRENT_PROVIDERS array",
  ],
]
`;

exports[`GuidedInstall AWS without awscli config shows the correct input prompts for collecting env vars 1`] = `
Array [
  Array [
    "Enter AWS account id:",
  ],
  Array [
    "Enter AWS account name:",
  ],
  Array [
    "Enter AWS region:",
  ],
  Array [
    "Enter AWS role name:",
  ],
  Array [
    "Enter Athena DB name:",
  ],
  Array [
    "Enter Athena DB table:",
  ],
  Array [
    "Enter Athena DB query result location (eg s3://your-athena-query-results-location):",
  ],
]
`;

exports[`GuidedInstall Azure shows the correct confirmation prompts for manual instructions 1`] = `
Array [
  Array [
    "Register a new Azure Application via your Azure Portal, under “App Registrations”.
	- You do not need to set a Redirect URI or configure platform Settings. Learn more about how to do this [here](https://docs.microsoft.com/en-us/azure/active-directory/develop/quickstart-register-app)",
  ],
  Array [
    "Within this application, go to “Certificates and secrets”, and create a new Client secret",
  ],
]
`;

exports[`GuidedInstall Azure shows the correct informational messages and documentation references 1`] = `
Array [
  Array [
    "Please follow the following prompts to configure Cloud Carbon Footprint to connect to a cloud provider. To connect to multiple cloud providers, please repeat this script for each additional provider. For an understanding of the default configurations and other configuration options, please visit the docs on the microsite: https://cloud-carbon-footprint.org/docs/introduction",
  ],
  Array [
    "Azure setup. See https://cloud-carbon-footprint.org/docs/azure for additional details.",
  ],
  Array [
    "Your Cloud Carbon Footprint application is now configured to use your cloud provider data. You can see these settings in the .env files listed above. To connect to an additional cloud provider or update an existing one, please re-run this script. For questions or more configuration options, please visit https://cloud-carbon-footprint.org/docs/introduction",
  ],
  Array [
    "Please make sure your appConfig object in the client Config file (packages/client/src/Config.ts) includes your cloud provider(s) in the CURRENT_PROVIDERS array",
  ],
]
`;

exports[`GuidedInstall Azure shows the correct input prompts for collecting env vars 1`] = `
Array [
  Array [
    "Enter Azure client id:",
  ],
  Array [
    "Enter Azure client secret:",
  ],
  Array [
    "Enter Azure tenant id:",
  ],
]
`;

exports[`GuidedInstall GCP shows the correct confirmation prompts for manual instructions 1`] = `
Array [
  Array [
    "Ensure you have a GCP Service Account with the permission to start BigQuery jobs and read Bigquery job results. Learn more about GCP Service Accounts [here](https://cloud.google.com/docs/authentication/getting-started).",
  ],
  Array [
    "Create and download a JSON private file for this Service Account to your local filesystem. Learn more about this authentication method [here](https://cloud.google.com/docs/authentication/getting-started)",
  ],
  Array [
    "Set up Google Cloud billing data to export to BigQuery. You can find the instructions for this [here](https://cloud.google.com/billing/docs/how-to/export-data-bigquery).",
  ],
]
`;

exports[`GuidedInstall GCP shows the correct informational messages and documentation references 1`] = `
Array [
  Array [
    "Please follow the following prompts to configure Cloud Carbon Footprint to connect to a cloud provider. To connect to multiple cloud providers, please repeat this script for each additional provider. For an understanding of the default configurations and other configuration options, please visit the docs on the microsite: https://cloud-carbon-footprint.org/docs/introduction",
  ],
  Array [
    "GCP setup. See https://cloud-carbon-footprint.org/docs/gcp for additional details.",
  ],
  Array [
    "Your Cloud Carbon Footprint application is now configured to use your cloud provider data. You can see these settings in the .env files listed above. To connect to an additional cloud provider or update an existing one, please re-run this script. For questions or more configuration options, please visit https://cloud-carbon-footprint.org/docs/introduction",
  ],
  Array [
    "Please make sure your appConfig object in the client Config file (packages/client/src/Config.ts) includes your cloud provider(s) in the CURRENT_PROVIDERS array",
  ],
]
`;

exports[`GuidedInstall GCP shows the correct input prompts for collecting env vars 1`] = `
Array [
  Array [
    "Enter the absolute path to your service account private key file:",
  ],
  Array [
    "Enter GCP account id:",
  ],
  Array [
    "Enter GCP account name:",
  ],
  Array [
    "Enter BigQuery table name:",
  ],
]
`;
